### PR TITLE
sysdata: over 9000 optimizations, 3 bugfixes

### PR DESCRIPTION
**Pull request by lasers**
Hi. Here to do the devil's work. ✝️ 

We clean up and optimize sysdata first before we tackle the number of the beast. https://github.com/ultrabug/py3status/issues/666 👿

* Chasing away the ghouls that feeds on dead bodies and the wastes in `C`, `F`, `K`.
* Consuming the sinful lies and the excessive lines.
* Checking units early and once in `post_config_hook` instead of loop.
* Calculating `multi(plier)` once in `post_config_hook` instead of loop.
* Carving out booleans in `post_config_hook` instead of using `py3.contain_formats()`.
* Changing things around to make things more organized and super readable.

Thx for looking. One day, we'll hit that `666` stars.

### CODE REVIEW

File: py3status/modules/sysdata.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
     swap_unit: the unit of swap to use in report, case insensitive.
         ['dynamic', 'KiB', 'MiB', 'GiB'] (default 'GiB')
     temp_unit: unit used for measuring the temperature ('C', 'F' or 'K')
```
Comment (lasers) [inline py3status/modules/sysdata.py]: [Satan's voice]
This kinda make more sense as `cpu_temp_unit` because it does not say which `temp` it is for... Could be `cpu`, `mem`, `ram`, or `???`. I wonder if we can obtain more with `psutil` or other tools. Although this is a bit long, it would be "correct" in the long run... We probably shouldn't change it anyway.

Alternatively, we could use `cpu_unit` to stand in line with other `*_unit`... assuming we are not going to expose CPU hertz units anytime soon... or will we? 👿 

Anyhow, I don't care. Up to you guys. Just pointing this out.

File: py3status/modules/sysdata.py [hunk 2]
```diff
@@ -3,18 +3,18 @@
 Display system RAM, SWAP and CPU utilization.
 
 Configuration parameters:
-    cache_timeout: how often we refresh this module in seconds (default 10)
-    format: output format string
+    cache_timeout: refresh interval for this module (default 10)
+    format: display format for this module
         *(default '[\?color=cpu CPU: {cpu_usage}%], '
         '[\?color=mem Mem: {mem_used}/{mem_total} GB ({mem_used_percent}%)]')*
     mem_unit: the unit of memory to use in report, case insensitive.
```
Comment (lasers) [inline py3status/modules/sysdata.py]: >case insensitive

Fake news 📰. I added `case insensitive`. I think we'd be better off sticking with `KiB`, `MiB`, `GiB`, `auto`.

**Comment by tobes** [comment]
Could you split this into PRs that make small changes?  You are reformatting lines/ renaming variables/changing documentation/changing defaults/changing quote types and changing the code all in one huge PR and it is too hard to read.

**Comment by lasers** [comment]
>Could you split this into PRs that make small changes?

I will do that. I'm going to deprecate (broken) `case insensitive` for quality assurance. This `case insensitive` feature likely got lost somewhere over time. One thing to consider... We need to remove deprecation stuffs now as there are too many and I wasn't able to use `substitute by value`... I hope is correct... to replace `dynamic` with `auto`... so I did it in `post_config_hook`.

>it is too hard to read.

Alternatively, I tell you when this is ready? One PR. Make your life easy. 👻
 
P.S. I want to work on my module (not window manipulation) and was wondering about https://github.com/ultrabug/py3status/issues/1059 as I want to continue and not losing motivation on this module... Do you want module name or keep it secret? I will try to split things up.  🎳

**Comment by lasers** [comment]
There are lot of PRs too. Just keep going? 😕

**Comment by tobes** [comment]
> I'm going to deprecate (broken) case insensitive

would not fixing it be a better idea?

**Comment by lasers** [comment]
Less lines this way... and it can be a little messy having to carry the values all the way through methods, returns, etc to where it need to be... Easier if we explicitly check bad units in `post_config_hook` and stop immediately..... Also, to support case insensitive, we probably would have to have a dict full of lowercase keys... so `kib`, `mib`, `gib` will show up in the logs (should we log them) instead of `KiB`, `MiB`, `GiB` (which I think is clean this way).

**Comment by tobes** [comment]
> kib

This is not a unit. 

Really we want to fix things properly which is going to need time and discussion imho

**Comment by lasers** [comment]
>>kib

>This is not a unit.

Then we could not support `case insensitive` because to have `kib` is to type `mem_unit = 'kib'` and to lowercase everything so we can match `kIB`, `KIB`, `gib`, `GIb`, etc... It's more clean to explicitly check for correct units than to make new dicts + `upper/lower` then carry them all the way.

If you looked at `self.multi`... that's in `post_config_hook`... calculated only once... then I carried them all the way to where it need to go... Not sure if it's worth doing that... or just keep calculating every loop in the same area. Wish we could measure modules. Sure on the discussion. Better not to support support lowercase units because way later in the methods.. we return `name[unit] * num`. (Bad unit).

I can remove them now then check things again... and to remove depreciation... and that could be done in a day.... but I can't be sure that this is the way to fix things properly for `psutils` and other tools. 

I just sort ABCs, organize things... gave them better naming, move things to `post_config_hook`, etc.

**Comment by lasers** [comment]
At this point, I'm optimizing the blood out of a stone for this module... that I see this...

```
        try:
            self.py3.threshold_get_color(
                max(cpu_usage, mem_used_percent), 'max_cpu_mem')
        except:
            try:
                self.py3.threshold_get_color(cpu_usage, 'max_cpu_mem')
            except:
                try:
                    self.py3.threshold_get_color(
                        mem_used_percent, 'max_cpu_mem')
                except:
                    pass
```
... could be something like this...
```
    if 'max_cpu_mem' in self.format:
        try:
            self.py3.threshold_get_color(
                max(cpu_usage, mem_used_percent), 'max_cpu_mem')
        except:
            try:
                self.py3.threshold_get_color(cpu_usage, 'max_cpu_mem')
            except:
                try:
                    self.py3.threshold_get_color(
                        mem_used_percent, 'max_cpu_mem')
                except:
                    pass
```
1) `if 'max_cpu_mem' in self.format:`
2) `self.py3.color_contains(self.format, 'max_cpu_mem')` _New!_
3) ... et cetera

Then we could skip that thresholds in loops... because `max_cpu_mem` is not part of default `format` and we don't have anything to capture `color=max_cpu_mem`.

Could `self.py3.threshold_get_color(load1, 'load')` internally check the formats first for a `load` in `load1`... then in the future loops, it either return something or immediately nothing without extra legwork? Similar to `format_contains` cache. Not sure if it's useful... All those `threshold_get_color` could just get smarter.

I'm really done with this. You still want separate PRs or discussion first?

**Comment by tobes** [comment]
>  I'm optimizing the blood out of a stone for this module

ok it sounds like this module is not where the work needs doing.  What it sounds like is needed is a clear understanding of what you are trying to achieve and why.  Messing around with the module feels like the wrong way to be looking at the issue.  The idea imho is to make writing/configuring modules as simple as possible by removing as much complication as we can and off loading it away from the modules.

> I'm really done with this. You still want separate PRs or discussion first?

I'm sort of lost on this.  What is it that you are trying to do?  There are too many changes to actually know what you're doing.  It also looks like the risk of breakage is  large and I'm not sure I see what advantages it gives.

I would prefer that we try to work out how to make things better in a slower more thought out way. an example would be around renaming a parameter, just a simple thing like that should have a reasoned case.

Also please do not do 
```
{cpu_temp}           cpu temperature                          eg 60
```

It causes future pain and is a unneeded maintenance drain.  I'm not sure the examples add much either

Anyhow currently we have a massive backlog of PRs and I suspect this will continue for a couple of weeks longer before improving

**Comment by lasers** [comment]
>Also please do not do

Done. Unaligned.

>ok it sounds like this module is not where the work needs doing.

It's basically finished. Just need to know what you'd like to do with the deprecation. Do we get rid of them all now? What about that `update_placeholder` thing? That percent PR we were talking about... etc.... 

> It also looks like the risk of breakage is large and I'm not sure I see what advantages it gives.

There shouldn't be any breakage, but no guarantee. Covering my ass. There were some things that were done loop......... was not necessary. They moved all to `post_config_hook`.... and just better naming, to make things more organized.... 

>a reasoned case.
I think `auto` would make more fitting name over `dynamic`. There are more... like is it worth keeping `temp_unit` / `cpu_temp_unit` as users could type them manually, but I think in this case, we might want to keep it... if this module were to grow much bigger..... and the percent thing... `cpu_usage` is a percent... why not `cpu_usage_percent` to follow those other `*percent}` placeholders? Really. It's the little things that we can worry about later... but I feel this is in much better condition now... and I'm getting myself familiar with this module... so I think I can add `htop` next if you want to merge this.

**Comment by tobes** [comment]
There is some useful stuff in here but it needs pulling into different parts.  I'm going to leave it to fester though as it has some useful bits among the noise

**Comment by ultrabug** [comment]
How do we want to move forward here? TBH @lasers there's too much for me to review simply, it's a mess but I'm sure you have strong things in this.

Split?

**Comment by lasers** [comment]
@ultrabug I rather let @tobes decide and if he want this to fester, we can let this fester. I plan to switch to Debian eventually. Meanwhile, I can test this again... but if both modules output looks same to you, then I believe I did a good job optimizing things.

I also have a working branch for #666 that depends on this.

IIRC The only change I made (despite the mess) is to put lot of things in `post_config_hook` than repeated tasks... and to remove case-insensitive docs. I wanted to remove this too:
```
    cpu_temp_unit: specify the cpu temperature unit to use
        ['C', '°C', 'F', '°F', 'K'] (default '°C')
```
to..
```
    cpu_temp_unit: specify the cpu temperature unit to use
        ['C', 'F', 'K'] (default '°C')
```
I can review this again, but there are other PRs I feel is ready.

**Comment by lasers** [comment]
Hi. I found an issue in this... with `['C', '°C', 'F', '°F', 'K']` for python2.7.
`[u'C', u'°C', u'F', u'°F', u'K']` fixed it, but I say `['F', u'C', 'K']` it and use `['C', 'F', 'K']` instead.

There are lot of things that could be done for this module... For one, I'd just like to send everything through the `thresholds` so the color names would be identical to placeholders and that users could have more access to color options... eg, individual `load1`, `load5`, `load15` thresholds and more... and we would miss nothing.

I think there are some things that were not exposed too... such as free, total, used, full... and maybe some in percents too... I think this kind of module is either fix everything we think is wrong with this or do easy life for all and just close this PR.

Deprecation like the thresholds function was messing with me because `thresholds = []` didn't work... and that we should make new specialized `thresholds` option... to default with...  since `thresholds = [(0, 'good'), (40, 'degraded'), (75, 'bad')]` wouldn't work well with some placeholders like `loads`, `cpu_temp`.

And other thing... `swap_unit` and `mem_unit` does not make sense... as users can select `['auto', 'KiB', 'MiB', 'GiB'] (default 'GiB')`... The `auto` (or former `dynamic`) determines this based on total amount instead of usage amount.... so if I have 8GB or 16GB... It will give me `GiB` and if you have 1GB ram netbook, you still get `GiB` too... We'll never get `KiB` with this... And the `swap` works the same way too. This basically calls for individual `auto` on every placeholder... which is probably tricky. We probably should see what we can do with `psutils` at this time.

Q. Should I should keep working on this? It gets harder (or annoying) to carry legacy stuffs... and that some things are just meh. Maybe it's a good idea to avoid math for this one so some things doesn't behave differently.... if we source everything from somewhere.

I dunno. Thoughts? I don't know how to read `0.01/3.73 GiB swap`. Looks like it's `14.5M`.

**Comment by lasers** [comment]
@ultrabug @tobes I'm exhausted from slaving away on this. Everything is working okay on my side and I can't find anything I'm missing. Please take it out for a spin. I didn't rename all placeholders because `temp` --> `temperature` is pretty long even although we always try to use full names.... and I don't know... It might be easier to write `psutil` module from scratch too.

Anyhow... Please let me know if I should proceed and rename bunch of them too... And if you want me to, I could continue and try to tackle #666 issue here too... Please! Please! Tell me if you ran into an issue... because I can't find anything no more.

I tested placeholders, color thresholds, old placeholders... This is ready to go in. `-137` is very nice.

**Comment by lasers** [comment]
I don't think I can do anything anymore for this. I'm all out of optimizations now. Review / merge.

I removed deprecation because it was messing with the results and I rounded the floats too. I wasn't able to do that in deprecation as some thing were not renamed and/or updated with the new formats... along with precision / padding.

I will try them again. See if we can get the floats to work too.

**Comment by lasers** [comment]
* I think we should remove `{cpu_temp}`, `{cpu_unit}` stuffs in favor of dedicated sensors module that does the job right so we can display the other sensors too.
* I think we should merge `net_rate.py` and `netdata.py` into one module `netdata` as both modules are parsing from the same file `/proc/net/dev` and having it separated is a bit confusing / silly (imo) so having a solid `netdata` would be nice. Simplifying is nice. Excluding the `simplify descriptions` PR, I hadn't touched `net_rate.py` yet.
* If after we merge `netdata` and `net_rate` together... Maybe it would be easier to make `cpudata` and `memdata` next... to address all those `memory` differences (eg, htop, top, free, etc...) all in one module along with `cpu` individual percents and whatnot.
* I wonder about `loads` because `i3status` already supports this. Maybe we could at least support the last two `load` placeholders too.
* Oh and `psutil`. Don't forget about `psutil`.

**Comment by lasers** [comment]
Questions about simplifying... and how much should I remove.

* I can remove `cpu_unit` `thermal_zone` (+ placeholders) in favor of a new module.
* I can remove `mem_unit`, `swap_unit` in favor of this newly added auto units.
* I wonder about `mem_htop = True`. Supposedly people prefer `mem_htop = True` over the current `mem` numbers, then we could switch it over and keep it that way. Maybe have a poll and see what people prefer as this is more about seeing if we can avoid making a new config in first place.
* Leaving us with only `cache_timeout` and `thresholds`.
* Keeping`mem_unit`, `swap_unit` there is now for fun because in the past, we have to decide which one to use... eg, for swap, probably `MiB` if we only use `swap_used` so we can see `315 MiB`... However with `swap total` in there too, it would be hard to parse, eg `8038 MiB`. I add `auto` units here so I want to know if we want to keep the configs or not.

**Comment by tobes** [comment]
@lasers shall we close this as we are now cleaning it up it neat chunks

**Comment by lasers** [comment]
I'll close this last after we finish every last details.
